discretisedfield/tools/tools.py

```diff
@@ -105,52 +105,59 @@ def topological_charge_density(field, /, method="continuous"):
         msg = f"Cannot compute topological charge density for {field.nvdim=} field."
         raise ValueError(msg)
 
-    if not field.mesh.attributes["isplane"]:
-        msg = (
-            "The field must be sliced before the topological "
-            "charge density can be computed."
+    if field.mesh.region.ndim != 2:
+        raise ValueError(
+            "The topological charge density can only be computed on fields with 2"
+            f" spatial dimensions, not {field.mesh.region.ndim=}."
         )
-        raise ValueError(msg)
 
     if method not in ["continuous", "berg-luescher"]:
         msg = "Method can be either continuous or berg-luescher"
         raise ValueError(msg)
 
-    axis1 = field.mesh.attributes["axis1"]
-    axis2 = field.mesh.attributes["axis2"]
     of = field.orientation  # unit field - orientation field
 
+    # Spatial axes associated with the first and second vector
+    # dimensions.
+    axis1 = field.vdim_mapping[field.vdims[0]]
+    axis2 = field.vdim_mapping[field.vdims[1]]
+
     if method == "continuous":
-        return (
-            1
-            / (4 * np.pi)
-            * of.dot(
-                of.diff(field.mesh.region.dims[axis1]).cross(
-                    of.diff(field.mesh.region.dims[axis2])
-                )
-            )
-        )
+        return 1 / (4 * np.pi) * of.dot(of.diff(axis1).cross(of.diff(axis2)))
 
     elif method == "berg-luescher":
+        axis1_idx = field.mesh.region._dim2index(axis1)
+        axis2_idx = field.mesh.region._dim2index(axis2)
+
         q = df.Field(field.mesh, nvdim=1)
 
         # Area of a single triangle
-        area = 0.5 * field.mesh.cell[axis1] * field.mesh.cell[axis2]
+        area = 0.5 * field.mesh.cell[axis1_idx] * field.mesh.cell[axis2_idx]
 
-        for i, j in itertools.product(range(of.mesh.n[axis1]), range(of.mesh.n[axis2])):
-            index = dfu.assemble_index(0, 3, {axis1: i, axis2: j})
+        for i, j in itertools.product(
+            range(of.mesh.n[axis1_idx]), range(of.mesh.n[axis2_idx])
+        ):
+            index = dfu.assemble_index(0, 2, {axis1_idx: i, axis2_idx: j})
             v0 = of.array[index]
 
             # Extract 4 neighbouring vectors (if they exist)
```

Comment:
We should probably also use valid in here? What do you think @samjrholt. If yes, I would prefer a separate PR for this.

Reply:
Yes you are correct, I have created #313.